---
title: "Introduction_TRskim"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction_TRskim}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction 

**TRskim** is an R package for tandem repeat visualization 

To download **TRskim**, use the following commands:

```r
require("devtools")
devtools::install_github("hassa963/TRskim", build_vignettes = TRUE)
library("TRskim")
```

To list all functions available in the package:

```r
ls("package:TRskim")
```
## Decompose Tandem Repeats

The function **decomposeTRs** decomposes the alleles into the given motifs and 
adds new motifs based onn the gaps between the motifs across the alleles.

```{r}
library(TRskim)
decomposed <- decomposeTRs(SOL1, motifs_SOL1) 

decomposed_TRs <- decomposed$compositions
motifs <- decomposed$motifs 

# SOL1 is a provided DNAStringSet of 95 Tandem repeat alleles for more details: 
# ?SOL1
# motifs_SOL1 is a provided DNAStringSet of 36 motif sequences within the SOL1 
# alleles for more details:
# ?motifs_SOL1

```

## Encode Tandem Repeats

The function **encodeTRs** encodes the decompose alleles into one character 
symbols which allows the alleles to be aligned based on its motifs not the 
individual bases. 

```{r}
encoded <- encodeTRs(decomposed_TRs, motifs)

encoded_TRs <- encoded$encoded
motif_maps <- encoded$map

```

## Align Tandem Repeats

The function **alignTRs** is a function that uses progressive alignment to align 
the tandem repeats my their motif composition

```{r}
aligned_TRs <- alignTRs(encoded_TRs)
```

## Plot Tandem Repeats 

The function **plotTR** is a function that uses the aligned tandem repeats to 
produce a tile plot visualization of the alleles 

```{r, fig.width=10, fig.height=6, out.width="100%"}
library(grid)
results <- plotTR(aligned_TRs, motif_maps, large = TRUE)

results$plot

grid.newpage()           # start a blank page
grid.draw(results$legend)

```

## Package reference 

- Hassan, N. (2025) TRskim: tandem repeat visualizer for BCB410H. Unpublished.
  https://github.com/hassa963/TRskim
  
## references 


- BioRender.com. BioRender [Online]. Available at: https://www.biorender.com
  (accessed 26 October 2025).
  
- OpenAI. ChatGPT (GPT-5) large language model (2025). https://chat.openai.com/

- Pagès, H., Aboyoun, P., Gentleman, R. & DebRoy, S. Biostrings: Efficient
  manipulation of biological strings (R package version 2.77.2, 2025).
  https://bioconductor.org/packages/Biostrings, doi:10.18129/B9.bioc.Biostrings

- Park, J., Kaufman, E., Valdmanis, P. N. & Bafna, V. TRviz: A Python Library
  for decomposing and Visualizing Tandem Repeat Sequences. Bioinformatics
  Advances 3, (2023).
  
- R Core Team. R: A Language and Environment for Statistical Computing. Vienna       Austria: R Foundation for Statistical Computing (2025). 
  https://www.R-project.org/
  
- Wickham, H. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag,
  New York, 2016

- Wickham, H. Reshaping data with the reshape package. J. Stat.
  Softw. 21, 1–20 (2007).
  
- Wilke, C. cowplot: Streamlined Plot Theme and Plot Annotations for ggplot2. R      package version 1.2.0, https://CRAN.R-project.org/package=cowplot (2025)

```{r}
sessionInfo()
```
